File: 323-number-of-connected-components-in-an-undirected-graph/323-number-of-connected-components-in-an-undirected-graph.scala
object Solution {
    def countComponents(n: Int, edges: Array[Array[Int]]): Int = {
        val id: Array[Int] = Array.fill(n)(0)

        for(i <- 0 to n - 1) { id(i) = i }

        def find(i: Int): Int = {
            if (id(i) != i) {
                val gi = find(id(i))
                id(i) = gi
            }
            id(i)
        }

        def union(a: Int, b: Int): Unit = {
            val ga = find(a)
            val gb = find(b)
            id(gb) = ga
        }

        edges.foreach(edge => union(edge(0), edge(1)))

        id.map(find(_)).toSet.size
    }
}

File: 323-number-of-connected-components-in-an-undirected-graph/323-number-of-connected-components-in-an-undirected-graph.java
class Solution {
    int[] group = null;
    
    public int countComponents(int n, int[][] edges) {
        group = new int[n];
        
        for(int i = 0; i < n; i++) {
            group[i] = i;
        }
        
        for(int[] edge : edges) {
            join(edge[0], edge[1]);
        }
        
        Set<Integer> res = new HashSet<>();
        
        for(int i = 0; i < n; i++) {
            res.add(findGroup(i));
        }
        
        return res.size();
    }
    
    int findGroup(int n) {
        if (group[n] != n) {
            int root = findGroup(group[n]);
            group[n] = root;
        }
        return group[n];
    }
    
    void join(int a, int b) {
        int grp1 = findGroup(a);
        int grp2 = findGroup(b);
        group[grp2] = grp1;
    }
}